lib/tty/pager/basic.rb

```diff
@@ -43,50 +45,71 @@ def default_prompt
       #
       # @api public
       def page(text, &callback)
-        page_num = 1
-        leftover = []
-        lines_left = @height
+        write(text, &callback)
+        reset
+      end
 
+      def write(text, &callback)
         text.lines.each do |line|
           chunk = []
-          if !leftover.empty?
-            chunk = leftover
-            leftover = []
+          unless @state.leftover.empty?
```

Comment:
I'm not sure about putting instance vars together in a state variable. It kind of makes parsing a bit 'too busy'.

Reply:
Yeah, I feel like one way to make this nicer would be to do a "law of demeter" pass and create sensible methods to query the state. For instance, `if @state.has_leftover?` or something. Would make this particular clause easier to read, since `unless empty` reads like a double-negative to me.

It was also a way to only check for the presence of a single variable to see if the pager has been started or not. But with the state of things right now, I don't think that would even be necessary. I'll try to turn `State` into a more useful abstraction, but if I don't manage it, I'm happy to roll the pieces back into their separate parts.

Reply:
I like the idea of `reset` and I'd probably keep all the mutating vars in there for now. In my experience, this simplifies the logic. I'd be all for extraction of the `State` if it attracted some common behaviour etc... In this instance, it feels unnecessary.

Reply:
Sorry, it may be a bit 'cheeky' of me to say but the fact that this is named `State` indicates that it doesn't have a clear role/purpose. For example, if it was named `PageView` or something similar, it would give a bit more context to its existence.

Reply:
It's a valid point -- the name "State" doesn't really encapsulate any domain logic, it's essentially a value struct that hold properties in one place.

I've pulled it out into a `Pagination` class that I think does a better job of encapsulating the domain logic of chunking the given lines into pages. It could even be initialized with the `@width` as well to handle the wrapping part, but I think it's fine like this. It's not reusable, but I think the separation makes the iteration and pagination a bit easier to read. It could also be tested separately if it were pulled out into another file, although I'm not sure if I'd go in that direction.

If you still think the separation adds a level of indirection that could be confusing or add overhead when reading, let me know and I'll revert to using the individual variables.

Reply:
I like what you've done here - the `Pagination` seems like more useful abstraction. Having said that, my thinking about this spike is to get the manual page chunks writing done first for all the pagers. From the point of what we're aiming to do this is a 'design improvement' diversion. I'd be inclined to skip it for now and first to do the `write` call as we want/need it. I think this spike has great stuff in it and I'd like to get to the point with it where I'm good to merge some implementation, whether you rebase later or submit another PR. Everything else can be done in separate PRs later. To me the smaller and focused PR the better. Plus I think that the basic pager line processing is a bit broken abstraction anyhow. It worked for me at the time. But now, it will fall down at the first bit when someone wants to page a continuous blob of text. So in other words, we may need to consider a different way of text processing for the basic pager anyhow. What do you think?